pkg/controller/poolcoordinator/podbinding/pod_binding_controller.go

```diff
@@ -0,0 +1,327 @@
+	nodeLister            listerv1.NodeLister
+	podSynced             cache.InformerSynced
+	podLister             listerv1.PodLister
+	getPodsAssignedToNode func(nodeName string) ([]*corev1.Pod, error)
+	podUpdateQueue        workqueue.RateLimitingInterface
+}
+
+func (c *Controller) onNodeUpdate(o interface{}, n interface{}) {
+	oldNode := o.(*corev1.Node)
+	newNode := n.(*corev1.Node)
+
+	// if node autonomy annotation changed, we need to handle all pods on this node except DaemonSet pod.
+	if oldNode.Annotations[projectinfo.GetAutonomyAnnotation()] != newNode.Annotations[projectinfo.GetAutonomyAnnotation()] {
+		pods, err := c.getPodsAssignedToNode(newNode.Name)
+		if err != nil {
+			klog.Errorf("failed to get pods for node(%s)", newNode.Name)
+			return
+		}
+
+		for i := range pods {
+			// skip DaemonSet pods and static pod
+			if isDaemonSetPodOrStaticPod(pods[i]) {
+				continue
+			}
+
+			// skip not running pods
+			if pods[i].Status.Phase != corev1.PodRunning {
+				continue
+			}
+
+			key, err := cache.MetaNamespaceKeyFunc(pods[i])
+			if err == nil {
+				c.podUpdateQueue.Add(key)
+				klog.Infof("pod(%s) tolerations should be handled for node annotation changed", key)
+			}
+		}
+	}
+}
+
+func (c *Controller) onPodUpdate(_ interface{}, n interface{}) {
+	newPod := n.(*corev1.Pod)
+
+	// skip DaemonSet pod and static pod
+	if isDaemonSetPodOrStaticPod(newPod) {
+		return
+	}
+
+	// skip not running pods
+	if newPod.Status.Phase != corev1.PodRunning {
+		return
+	}
+
+	needHandled := false
+	// only handle pods with binding annotation or work on node with autonomy annotation
+	if newPod.Annotations != nil && newPod.Annotations[constant.PodBindingAnnotation] == "true" {
+		needHandled = true
+	} else if len(newPod.Spec.NodeName) != 0 {
+		node, err := c.nodeLister.Get(newPod.Spec.NodeName)
+		if err != nil {
+			klog.Errorf("failed to get node(%s) for pod(%s/%s)", newPod.Spec.NodeName, newPod.Namespace, newPod.Name)
+			return
+		}
+
+		if node.Annotations != nil && node.Annotations[projectinfo.GetAutonomyAnnotation()] == "true" {
+			needHandled = true
+		}
+	}
+
+	if needHandled {
+		key, err := cache.MetaNamespaceKeyFunc(newPod)
+		if err == nil {
+			klog.Infof("pod(%s) tolerations should be handled for pod update", key)
+			c.podUpdateQueue.Add(key)
+		}
+	}
+}
+
+func NewController(kc client.Interface, informerFactory informers.SharedInformerFactory) *Controller {
+	ctl := &Controller{
+		client:         kc,
+		podUpdateQueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	nodeInformer := informerFactory.Core().V1().Nodes()
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: ctl.onNodeUpdate,
+	})
+	ctl.nodeSynced = nodeInformer.Informer().HasSynced
+	ctl.nodeLister = nodeInformer.Lister()
+
+	podInformer := informerFactory.Core().V1().Pods()
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: ctl.onPodUpdate,
```

Comment:
Do we need handle add event?

Reply:
we need to wait pod phase become `Running` before changing tolerations for pod, so it doesn't need to handle add event.